### DISCUSSION TITLE
Error: Unable to snapshot resolve dependencies

### DISCUSSION THREAD

**Discussion by Smanst3r** [webpack/webpack#16417] [Q&A] [answered]
# Bug report
In process of building "filesystem" cache webpack hits the /home and root / directories and falls with EACCES error 
PackFileCacheStrategy :: Error: Unable to snapshot resolve dependencies
![Screenshot from 2022-10-26 12-07-59](https://user-images.githubusercontent.com/8392953/197988209-4a40c681-bbb9-4a18-986c-6375e2c6b516.png)

**What is the current behavior?**
Current behavior: function `_readFileHash` in file **lib/FileSystemInfo.js** falls on error EACCES. 
Because the `path` parameter has values: "/home", "/".
See **FileSystemInfo.js**:
![Screenshot from 2022-10-26 12-07-59](https://user-images.githubusercontent.com/8392953/197985318-e0261e93-85ce-4fef-a867-617fa7b9faaa.png)

And system configuration does not allow to access the folders above project directory. 
The question is: why webpack is trying to resolve such paths: "/home", "/" and how to prevent this  
The `context` parameter does not help 

**If the current behavior is a bug, please provide the steps to reproduce.**
- Set 
`cache: {
    type: 'filesystem',
    buildDependencies: {
        // This makes all dependencies of this file - build dependencies
        config: [__filename],
        // By default webpack and loaders are build dependencies
    },
}`

**What is the expected behavior?**
Prohibit reading of files above the project directory

**Other relevant information:**
webpack version: 5.74.0
Node.js version: v16.17.1
Operating System: Linux
Additional tools: webpack config generated via Symfony encore, but for plain config the result is the same

**Comment by alexander-akait** [comment]
> In process of building "filesystem" cache webpack hits the /home and root / directories and falls with EACCES error
PackFileCacheStrategy :: Error: Unable to snapshot resolve dependencies

It can happen when you have `import foo from "/test.js"`, i.e. absolute URLs, also resolver allows to change behaviour and set for example `modules` to root and webpack again will do it, if you need the real answer we need your configuration as minimum

**Reply by Smanst3r** [reply]
Hi @alexander-akait . Thanks for reply. 
Configuration sample to reproduce webpack hits "/home" and "/" directories issue.

**webpack.config.test.js**
<pre>
const path = require('path');

module.exports = {
    mode: 'production',
    output: {
        path: path.resolve('../public_html/static/compiled/'),
        publicPath: '/static/compiled/',
    },
    entry: {
        'app': [
            './src/AppBundle/Resources/assets/js/website-search.js',
        ]
    },
    cache: {
        type: 'filesystem',
        buildDependencies: {
            // This makes all dependencies of this file - build dependencies
            config: [__filename],
            // By default webpack and loaders are build dependencies
        },
    },
    resolve: {
        extensions: ['', '.js', '.jsx'],
    },
    module: {
        rules: [
            {
                test: /\.m?js$/,
                exclude: /(node_modules|bower_components)/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        presets: ['@babel/preset-env']
                    }
                }
            },
            {
                test: /\.css$/i,
                use: ["style-loader", "css-loader"],
            },
        ]
    }
};
</pre>

There is output of `./node_modules/webpack/bin/webpack.js --config ./webpack.config.test.js`
[_readFileHash(just console.log output).txt](https://github.com/webpack/webpack/files/9868904/_readFileHash.just.console.log.output.txt)

**Comment by gonandriy** [comment]
The reason behind that is 
https://github.com/webpack/enhanced-resolve/blob/main/lib/SymlinkPlugin.js
```
if (resolveContext.fileDependencies)
    resolveContext.fileDependencies.add(path);
```
fills _resolveContext.fileDependencies_ with array containing paths to very top of filesystem tree.
It looks like it is done deliberately, as there are a test case for that
https://github.com/webpack/enhanced-resolve/blob/main/test/dependencies.js

The bug is reproduce-able when you restrict the **/home** directory access for users, so it would look like that
`ls -lZ /`
`drwx--x--x. root root system_u:object_r:home_root_t:s0 home`

Not sure how to fix that and is it worth to..

Regards,

**Comment by Smanst3r** [comment, marked as answer]
Resolved this issue by adding **snapshot** option to config:
`.snapshot {
    immutablePaths: [/^\/home$/, /^\/$/],
}`
